cpp/oneapi/dal/detail/array_impl.hpp

```diff
@@ -236,7 +214,15 @@ class array_impl : public base {
 #endif
     }
 
-    detail::shared<T> copy() {
+    template <typename Y>
+    void check_array_has_ownership_structure(const array_impl<Y>& ref) {
+        if (ref.get_if_cshared() == nullptr && ref.get_if_shared() == nullptr) {
```

Comment:
This will never happen, `variant<T, U>`  always contains either instance of `T` or `U`. There are some exceptions, see https://en.cppreference.com/w/cpp/utility/variant/valueless_by_exception

However, the described situation happen only if `T` or `U` raises exception during changing the `variant` state. This will never happen for `shared_ptr`, because all copy/move constructors/assignment are noexcept.

So this check is useless.

Reply:
I'd recommend put assertion `ONEDAL_ASSERT(!data_owned_.valueless_by_exception())`